consensus/polybft/slashing/double_signing_tracker.go

```diff
@@ -0,0 +1,189 @@
+
+	"github.com/0xPolygon/polygon-edge/consensus/polybft/wallet"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+type SenderMessagesMap map[types.Address][]*ibftProto.Message
+
+type DoubleSignEvidence struct {
+	signer   types.Address
+	round    uint64
+	messages []*ibftProto.Message
+}
+
+func newDoubleSignEvidence(signer types.Address, round uint64,
+	messages []*ibftProto.Message) *DoubleSignEvidence {
+	return &DoubleSignEvidence{signer: signer, round: round, messages: messages}
+}
+
+type MessagesMap map[uint64]map[uint64]SenderMessagesMap
+
+// getSenderMsgs returns commit messages for given height, round and sender
+func (m MessagesMap) getSenderMsgs(view *ibftProto.View, sender types.Address) []*ibftProto.Message {
+	if roundMsgs, ok := m[view.Height]; ok {
+		if senderMsgsMap, ok := roundMsgs[view.Round]; ok {
+			return senderMsgsMap[sender]
+		}
+	}
+
+	return nil
+}
+
+// registerSenderMsgs registers messages for the given height, round and sender
+func (m MessagesMap) registerSenderMsgs(view *ibftProto.View, sender types.Address, msgs []*ibftProto.Message) {
+	var senderMsgs SenderMessagesMap
+
+	if roundMsgs, ok := m[view.Height]; !ok {
+		roundMsgs = make(map[uint64]SenderMessagesMap)
+		m[view.Height] = roundMsgs
+		senderMsgs = createSenderMsgsMap(view.Round, roundMsgs)
+	} else if senderMsgs, ok = roundMsgs[view.Round]; !ok {
+		senderMsgs = createSenderMsgsMap(view.Round, roundMsgs)
+	}
+
+	senderMsgs[sender] = msgs
+}
+
+// createSenderMsgsMap initializes senders message map for the given round
+func createSenderMsgsMap(round uint64, roundMsgs map[uint64]SenderMessagesMap) SenderMessagesMap {
+	sendersMsgs := make(SenderMessagesMap)
+	roundMsgs[round] = sendersMsgs
+
+	return sendersMsgs
+}
+
+type DoubleSigningTracker interface {
+	Handle(msg *ibftProto.Message)
+	GetEvidences(height uint64) []*DoubleSignEvidence
+	PruneMsgsUntil(height uint64)
+}
+
+type DoubleSigningTrackerImpl struct {
+	preprepare  MessagesMap
+	prepare     MessagesMap
+	commit      MessagesMap
+	roundChange MessagesMap
+
+	logger hclog.Logger
+}
+
+func NewDoubleSigningTracker(logger hclog.Logger) *DoubleSigningTrackerImpl {
+	return &DoubleSigningTrackerImpl{
+		logger:      logger,
+		preprepare:  make(MessagesMap),
+		prepare:     make(MessagesMap),
+		commit:      make(MessagesMap),
+		roundChange: make(MessagesMap),
+	}
+}
+
+// Handle is implementation of IBFTMessageHandler interface, which handles IBFT consensus messages
+func (t *DoubleSigningTrackerImpl) Handle(msg *ibftProto.Message) {
+	signer, err := wallet.RecoverSignerFromIBFTMessage(msg)
+	if err != nil {
+		t.logger.Warn("failed to recover signer upon message receive",
+			"message", msg)
+
+		return
+	}
+
+	sender := types.BytesToAddress(msg.From)
+	// ignore messages where signer and sender are not the same
+	if signer != sender {
+		t.logger.Warn("signer and sender of IBFT message are not the same, ignoring it...",
+			"message", msg)
+
+		return
+	}
```

Comment:
We should also check if the sender is an active validator.

Reply:
This is still to be discussed since a validator could have a stake greater than 0, but it could happen that he is not elected for the current epoch validator (sub)set if the max validator set size is already reached.

Reply:
Fixed by #1767.